### ISSUE TITLE
Readahead does not support non-blocking read

### ISSUE THREAD

**Issue by git-blame** [klauspost/readahead#20]
I know readahead does not make this claim. But the fact that it is async in a separate goroutine and readme states`splitting an input stream into concurrent processing, and also helps smooth out bursts of input or output` made me think it might. However, wrapping readahead reader with with a bufio reader shows that readahead blocks in `fill()`.

```golang
		if a.cur != nil {
			a.reuse <- a.cur
			a.cur = nil
		}
		b, ok := <-a.ready // blocks here
		if !ok {

```

Interestingly, bufio does handle 0-byte read from the wrapped reader and will simply return `io.ErrNoProgress`. So it does have a pseudo/synthetic non-blocking read if the inner reader returns empty buffers.
```golang
func (b *Reader) fill() {
...
    for i := maxConsecutiveEmptyReads; i > 0; i-- {
        n, err := b.rd.Read(b.buf[b.w:])
        if n < 0 {
            panic(errNegativeRead)
        }
        b.w += n
        if err != nil {
            b.err = err
            return
        }
        if n > 0 {
            return
        }
    }
    b.err = io.ErrNoProgress
```
```

**Comment by klauspost** [comment]
Consider [this package](https://github.com/smallnest/ringbuffer). It offers both blocking and non-blocking reads.

Mostly I have replaced my own usage with that package, which I helped by adding some features.

**Comment by klauspost** [comment]
(closing, since I don't really have any plans to add new features here)

**Comment by git-blame** [comment]
Thanks, I'll look at that package. Actually, this package doesn't seem to work well with intermittent data like a pipe. So I wrapped a reader around the pipe which simply reports the number of bytes read. Then I passed the reader to readahead.

My reader reports that X number of bytes have been read. But if I do a read with a large buffer > X on readahead, instead of simply returning X, readahead blocks inside the same location in function `fill()`.
